### PR TITLE
8330647: Two CDS tests fail with -UseCompressedOops and UseSerialGC/UseParallelGC

### PR DESCRIPTION
**Pull request by calvinccheung**
The following two tests contain `@requires vm.cds.write.archived.java.heap` which implies `UseG1GC && UseCompressedClassPointers`
```
cds/appcds/cacheObject/ArchiveHeapTestClass.java
cds/serviceability/ReplaceCriticalClassesForSubgraphs.java
```
The tests would fail if options conflicting with `vm.cds.write.archived.java.heap` are specified via the ` -Dtest.cds.runtime.options` property because the options would be added after the `@require` check.

A fix is to check if the ` -Dtest.cds.runtime.options` property contains non-null value and throws a `SkippedException`.

Update:
The checking of the ` -Dtest.cds.runtime.options` property is now performed in VMProps.java.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330647](https://bugs.openjdk.org/browse/JDK-8330647): Two CDS tests fail with -UseCompressedOops and UseSerialGC/UseParallelGC (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [83529521](https://git.openjdk.org/jdk/pull/19274/files/83529521a009945d83b38a9e61e1b81d3650b0e3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19274/head:pull/19274` \
`$ git checkout pull/19274`

Update a local copy of the PR: \
`$ git checkout pull/19274` \
`$ git pull https://git.openjdk.org/jdk.git pull/19274/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19274`

View PR using the GUI difftool: \
`$ git pr show -t 19274`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19274.diff">https://git.openjdk.org/jdk/pull/19274.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19274#issuecomment-2116286592)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Can't the code that implements the `@requires` check also check for that property?

**Comment by calvinccheung** [comment]
> Can't the code that implements the `@requires` check also check for that property?

Yes, I've reverted the changes in the tests and added the checking of the `test.cds.runtime.options` property in `VMProps.java`.
Also, instead of just checking for null value, it checks for if any of the options conflicts is not compatible with writing Java heap objects into the CDS archive.

File: test/jtreg-ext/requires/VMProps.java [hunk 1]
```diff
@@ -462,7 +462,29 @@ protected String vmCDSForCustomLoaders() {
...
+        if (jtropts == null)
+            return true;
+        String CCP_DISABLED = "-XX:-UseCompressedClassPointers";
+        for (String opt : jtropts.split(",")) {
+            if (opt.equals(CCP_DISABLED))
+                return false;
+            if (opt.startsWith(GC_PREFIX) && opt.endsWith(GC_SUFFIX)) {
+                String gc = opt.substring(GC_PREFIX.length(), opt.length() - GC_SUFFIX.length());
+                if (!gc.equals("G1"))
+                    return false;
```
Comment (dholmes-ora) [inline test/jtreg-ext/requires/VMProps.java:484]: Can't you just check for `-XX:+UseG1GC` ?

**Review by dholmes-ora — COMMENTED** [review]
Update looks good. One query below. Thanks

**Review by iklam — APPROVED** [review]
LGTM.

File: test/jtreg-ext/requires/VMProps.java [hunk 2]
```diff
@@ -333,6 +333,8 @@ protected void vmGC(SafeMap map) {
         map.put("vm.gc.ZSinglegen", () -> "" + (vmGCZ && (!genZ || genZIsDefault)));
     }
 
+    final String GC_PREFIX  = "-XX:+Use";
+    final String GC_SUFFIX  = "GC";
```
Comment (turbanoff) [inline test/jtreg-ext/requires/VMProps.java:337]:
```suggestion
    final String GC_PREFIX = "-XX:+Use";
    final String GC_SUFFIX = "GC";
```

File: test/jtreg-ext/requires/VMProps.java (hunk 1 above)
Comment (calvinccheung) [reply test/jtreg-ext/requires/VMProps.java]: I've pushed another commit based on your suggestion.

File: test/jtreg-ext/requires/VMProps.java (hunk 2 above)
Comment (calvinccheung) [reply test/jtreg-ext/requires/VMProps.java]: Fixed.

File: test/jtreg-ext/requires/VMProps.java [hunk 3]
```diff
@@ -462,7 +462,28 @@ protected String vmCDSForCustomLoaders() {
...
+        if (jtropts == null)
+            return true;
+        String CCP_DISABLED = "-XX:-UseCompressedClassPointers";
+        String G1GC_ENABLED = "-XX:+UseG1GC";
+        for (String opt : jtropts.split(",")) {
+            if (opt.equals(CCP_DISABLED))
+                return false;
+            if (opt.startsWith(GC_PREFIX) && opt.endsWith(GC_SUFFIX) &&
+                !opt.equals(G1GC_ENABLED))
+                return false;
```
Comment (dholmes-ora) [inline test/jtreg-ext/requires/VMProps.java:484]: Based on the bug report both of these conditions are needed to be enabled for there to be a problem, but you exclude based on either `-XX:-UseCompressedClassPointers"` or any non-G1 GC. ??

Comment (calvinccheung) [reply test/jtreg-ext/requires/VMProps.java:484]: I've checked again, without the fix, the test fails if running with either `-Dtest.cds.runtime.options=-XX:-UseCompressedClassPointers` or any non-G1GC such as `-Dtest.cds.runtime.options=-XX:+UseZGC`.

File: test/jtreg-ext/requires/VMProps.java [hunk 4]
```diff
@@ -333,6 +333,8 @@ protected void vmGC(SafeMap map) {
         map.put("vm.gc.ZSinglegen", () -> "" + (vmGCZ && (!genZ || genZIsDefault)));
     }
 
+    final String GC_PREFIX = "-XX:+Use";
+    final String GC_SUFFIX = "GC";
```
Comment (stefank) [inline test/jtreg-ext/requires/VMProps.java:337]: Maybe move them to the beginning of the file and make them `private static final`? At least add a blank line after them and the following comment.

File: test/jtreg-ext/requires/VMProps.java [hunk 5]
```diff
@@ -462,7 +462,28 @@ protected String vmCDSForCustomLoaders() {
...
-        return "" + ("true".equals(vmCDS()) && WB.canWriteJavaHeapArchive());
+        return "" + ("true".equals(vmCDS()) && WB.canWriteJavaHeapArchive()
+                     && isCDSRuntimeOptionsCompatible());
+    }
+
+    /**
+     * @return true if the VM options specified via the "test.cds.runtime.options"
+     * property is compatible with writing Java heap objects into the CDS archive
+     */
+    protected boolean isCDSRuntimeOptionsCompatible() {
```
Comment (stefank) [inline test/jtreg-ext/requires/VMProps.java:473]: Please add braces for all the added `if` statements.

**Review by stefank — CHANGES_REQUESTED** [review]
This is not a review of the functional changes of this fix, but since this patch does change the shared VMProps.java file I have some style requests.

File: test/jtreg-ext/requires/VMProps.java (hunk 4 above)
Comment (calvinccheung) [reply test/jtreg-ext/requires/VMProps.java]: I've moved them to the beginning of the file.

File: test/jtreg-ext/requires/VMProps.java (hunk 5 above)
Comment (calvinccheung) [reply test/jtreg-ext/requires/VMProps.java:474]: Fixed.

Comment (stefank) [reply test/jtreg-ext/requires/VMProps.java:474]: You missed this one:
```
        if (jtropts == null)
            return true;
```

**Review by stefank — CHANGES_REQUESTED** [review]

File: test/jtreg-ext/requires/VMProps.java (hunk 5 above)
Comment (calvinccheung) [reply test/jtreg-ext/requires/VMProps.java:474]: Fixed.

**Review by dholmes-ora — APPROVED** [review]
Okay that seems fine if either condition disables writing into the CDS archive.

Thanks

**Comment by calvinccheung** [comment]
Thanks @iklam @dholmes-ora @stefank @turbanoff for the review.

/integrate

**Pull request closed without merging**